### DISCUSSION TITLE
How do I implement comments in Lexical?

### DISCUSSION THREAD

**Discussion by stuburger** [facebook/lexical#1839] [Q&A] [answered]
A common requirement in many text editors is the ability to add and reply to comments/annotations that apply to a range of text.  What is the best approach to do this using Lexical? Would I be extending `TextNode` or `ElementNode`? **It would be neat if the Lexical docs provided an example of how one might implement comment functionality.** 🙌 

## Some background: implementing comments in Prosemirror

In Prosemirror you can implement comments in one of two ways:
1. **using a data structure** that keeps track of start/end positions of comments (recommended)
2. or **using marks** which makes comments part of the prosemirror doc which is part of the editorState

Both approaches have their pros and cons. You can find a number of threads discussing the tradeoffs. **I am also interested to know if the same tradeoffs exist in Lexical?**?

[Here](https://discuss.prosemirror.net/t/inlining-a-node-for-a-comment-plugin-or-best-to-use-marks/2391/6) is a very good thread that talks about these two approaches and the tradeoffs.

### Comments as Decorations:
* decouples comments from the editor state
  * comments live inside a separate data structure
  * easier to work with (no Prosemirror APIs)
  * treats comments as a view concern
* store `from` `to` values and a `comment_id`
* must roll your own to get history and copy+paste behaviour if you need it.
* remains performant while typing when there are thousands of comments

### Comments as Marks:
* comments live inside the main ProseMirror document
* store only `comment_id` on mark.attrs (i.e. `from` and `to` positions are implied by the marks)
* works with history and copy+paste out of the box
* comments spanning multiple block nodes (e.g. paragraphs) require more than one mark
* harder to access (must use Prosemirror's API to access them: e.g.`doc.descendants`, etc)
* typing performance suffers when you have thousands of comment marks in the document as marks are always serialised as dom nodes

I'm curious to know if similar tradeoffs exist in Lexical or if there is a more opinionated way to achieve comment functionality. 

Thanks for open sourcing Lexical. Even with incomplete docs I've found Lexical easier to pick up than Prosemirror (which without exaggeration took me weeks to start feeling comfortable with). Great work!

**Comment by zurfyx** [comment]
Exciting topic! @trueadm and @sereneinserenade are also exploring this space right now and there's some good discussion on the [showcase group on Discord](https://discord.com/channels/953974421008293909/964928369793843230/965332351624294440) if you want to join the conversation. Later, we can maybe summarize the conversation here for historical purposes.

**Comment by sereneinserenade** [comment, marked as answer]
@stuburger Just for the reference, here's the Commenting-Solution I created using `ElementNode` inspired by `LinkNode` from Lexical.

repo: https://github.com/sereneinserenade/lexical-comments
demo: https://sereneinserenade.github.io/lexical-comments/

**Reply by stuburger** [reply]
Its fantastic to see that folks are thinking hard about this problem already. @zurfyx has a message on discord that pretty much covers everything I've summed up in my question about the different approaches. I'm stoked everyone seems to be on the same page about this.

@sereneinserenade thank you for the links and for breaking ground here so quickly! If I contrast your approach to a Prosemirror implementation, your take on comments in Lexical its similar to using marks in that the comments are stored in the document itself.  Would you say thats accurate? I've bookmarked your repo as its a great reference to share with my team. I am still curious about the "separate data structure" approach (are there names for these different approaches I wonder? 🤔 ) 

There is also mention of an implementation that extends `DecorationNode` to implement comments - I look forward to seeing it. We have a use case where we may have A LOT of comments (often thousands) and the editor becomes unusable using the ElementNode approach. I tested this in the Lexical playground by duplicating a bunch of LinkNodes and then typing in the editor. You'll notice that there is severe input lag. I figured that'd be a good way to get a ballpark understanding of what input performance feels like with lots of these inline ElementNodes. 

I figure the way that @trueadm is looking at building using DecorationNodes will hopefully be able to make better use of reconciler optimizations as it keeps Lexical nodes flat. Very curious to see what comes of that work!

**Reply by trueadm** [reply]
From discord:

So digging into this more. There are a ton of caveats around this. Earlier this week I added "marks" to TextNodes, that allowed us to mark values on node as offsets. This could then be used to associate data externally. However, this has problems of its own. It means we then need to absolutely position DOM elements above the text – like we do in collaboration. It's not so bad in collaboration as the cursors and selection can be debounced and throttled if there are too many on at once – given the heavy use of getBoundingClientRect.

If we go down this path, then we make it expensive for drawing when lots of editing occurs at once. It's actually far better to update the EditorState and wrap nodes in an Element type of Mark. However, if you do that, then you can't have overlapping marks. So we'd need to make a base Mark element in core that did this and would handle overlapping.

Another issue is that there are bunch of a11y semantics around using the DOM you get, and you lose if you overlap selection markers on top. Checking in with our a11y team and this is indeed a violation, we should be using <mark> elements, as someone brought up in Q&A.

I'll do some more digging!

**Reply by strdr4605** [reply]
@trueadm, any conclusion on this?
Did you manage to do:
> I was planning on building a comments implementation next week that works slightly different

What about suggestions?

While discovering remirror, I thought that suggestions should be similar to annotations, but Google Docs has a bit different experience. When doing suggestions you enter a different mode, more like doing a git patch to the doc. (should this be a different discussion🤔)

---
P.S.
My team plans to add support for annotations and suggestions in Q4. Now we are focused on WYSIWYG and collaboration.
While researching the solutions (in Q1) we did not consider lexical (closed source and premature at that time).
Now I am wondering if we should give it a try considering that we need:
- [x] WYSIWYG
- [x] collaborative
- [x] annotations
- [x] images
- [x] links, hashtags, mentions 
- [ ] suggestions
- [ ] emojis 
- [ ] React Native support (Webview or Native)
🤔😅

**Reply by trueadm** [reply]
The playground has commenting, if you check it out!

**Reply by strdr4605** [reply]
Do you plan to release it as a separate package? Or it's still WIP?

**Reply by ChannKan** [reply]
@sereneinserenade 
Does your approach for comments in lexical work also if there is 2 lexical editors on the same page?

**Reply by sereneinserenade** [reply]
Hey @ChannKan please use lexical's inbuilt solution for comments, my solution isn't updated in a while. Thanks

**Comment by chrisrickard** [comment]
@trueadm Awesome, is the commenting code in the playground available online?

**Reply by trueadm** [reply]
Yep. If you checkout the playground you can make inline comments when selecting some content or just post comments for the document

**Comment by chrisrickard** [comment]
Yeah cool, is the sourcecode to the playground available into repo?

On Mon, 19 Dec 2022 at 8:57 pm, Dominic Gannaway ***@***.***>
wrote:

> Yep. If you checkout the playground you can make inline comments when
> selecting some content or just post comments for the document
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/facebook/lexical/discussions/1839#discussioncomment-4447349>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEEYCMI4Z2VBA3J7ABJD43WOAWRDANCNFSM5TVNF5BA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
Chris Rickard

**Reply by trueadm** [reply]
Yep it’s all in the GitHub repository you’re viewing. Just look for the lexical-playground package

**Comment by chrisrickard** [comment]
Thanks!

On Tue, Dec 20, 2022 at 2:39 AM Dominic Gannaway ***@***.***>
wrote:

> Yep it’s all in the GitHub repository you’re viewing. Just look for the
> lexical-playground package
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/facebook/lexical/discussions/1839#discussioncomment-4450074>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEEYCPBHDA225QL6GY2PGDWOB6STANCNFSM5TVNF5BA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Chris Rickard

**Comment by erickjth** [comment]
Hey @trueadm , I have been playing with the commenting code from the playground, and it works great, but I have a use case where I need to persist these comments and restore them on loading the page, so I have been trying to do by myself checking the lexical-yjs, more specifically this file [SyncEditorStates.ts](https://github.com/facebook/lexical/blob/main/packages/lexical-yjs/src/SyncEditorStates.ts) to follow how to restore the selection and then run the `$wrapSelectionInMarkNode` to restore the marks, but my attempts have been unsuccessful.  Could you please help me with what I need to save and what I need to use to restore those?

**Reply by sonu-44** [reply]
Hi @erickjth  , Have you found any solution for that , I am struggling with the same thing ?

**Reply by erickjth** [reply]
Hi @sonu-44. Yeah, kind of. Somebody advised me to grab the anchorNode/focusNode paths, plus the offsets. I created this gist you can try to follow. Hopefully it can help.

https://gist.github.com/erickjth/3946e4e21b4d2f6917afd877a01bb38f

**Reply by sep2** [reply]
This doesn't seem to work if you have two marks overlapping the same TextNode, because [this line](https://gist.github.com/erickjth/3946e4e21b4d2f6917afd877a01bb38f#file-gistfile1-ts-L89) modifies the TextNode therefore the second mark can't find it by the original anchorNode/focusNode path anymore. I tried to create all the selections upfront but still no luck.
